modules/ml/src/main/java/org/apache/ignite/ml/trainers/group/GroupTrainer.java

```diff
@@ -0,0 +1,206 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.trainers.group;
+
+import java.io.Serializable;
+import java.util.UUID;
+import java.util.stream.Stream;
+import org.apache.ignite.Ignite;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.ml.Model;
+import org.apache.ignite.ml.math.functions.IgniteBinaryOperator;
+import org.apache.ignite.ml.math.functions.IgniteFunction;
+import org.apache.ignite.ml.math.functions.IgniteSupplier;
+import org.apache.ignite.ml.trainers.Trainer;
+import org.apache.ignite.ml.trainers.group.chain.ComputationsChain;
+import org.apache.ignite.ml.trainers.group.chain.EntryAndContext;
+import org.apache.ignite.ml.trainers.group.chain.HasTrainingUUID;
+
+/**
+ * Class encapsulating synchronous distributed group training.
+ * Training is performed by following scheme:
+ * 1. For specified set of keys distributed initialization is done. For each key some initialization result is returned.
+ * 2. All initialization results are processed locally and reduced into some object of type I.
+ * 3. While 'shouldContinue' condition is true, training loop step is executed.
+ * 4. After loop is finished, data from each key from final key set is collected.
+ * 5. Data collected on previous step is transformed into a model which is returned as final result.
+ * Note that all methods returning functions, suppliers etc should return values with minimal dependencies because they are serialized
+ * with all dependent objects.
+ *
+ * @param <LC> Type of local context of the training.
+ * @param <K> Type of cache keys on which the training is done.
+ * @param <V> Type of cache values on which the training is done.
+ * @param <IN> Type of data returned after initializing of distributed context.
+ * @param <R> Type of result returned after training from each node.
+ * @param <I> Type of data which is fed into each training loop step and returned from it.
+ * @param <M> Type of model returned after training.
+ * @param <T> Type of input to this trainer.
+ * @param <G> Type of distributed context which is needed for forming final result which is send from each node to trainer for final model creation.
+ */
+abstract class GroupTrainer<LC extends HasTrainingUUID, K, V, IN extends Serializable, R extends Serializable, I extends Serializable, M extends Model, T extends GroupTrainerInput<K>, G> implements Trainer<M, T> {
+    /**
+     * Cache on which training is performed. For example it can be cache of neural networks.
+     */
+    IgniteCache<GroupTrainerCacheKey<K>, V> cache;
+
+    /**
+     * Ignite instance.
+     */
+    Ignite ignite;
+
+    /**
+     * Construct an instance of this class.
+     *
+     * @param cache Cache on which training is performed.
+     * @param ignite Ignite instance.
+     */
+    public GroupTrainer(
```

Comment:
given that class is declared package private, public modifier here looks redundant. Consider either declaring class public or removing public modifier from constructor